src/supa/job/reserve.py

```diff
@@ -448,20 +489,197 @@ def __init__(self, connection_id: UUID):
+        If the reservation state machine is not in the correct state for a ReserveCommit
+        an NSI error is returned leaving the state machine unchanged.
+        """
+        self.log.info("Aborting reservation")
+
+        from supa.db.session import db_session
+
+        with db_session() as session:
+            reservation = (
+                session.query(Reservation).filter(Reservation.connection_id == self.connection_id).one_or_none()
+            )
+            rsm = ReservationStateMachine(reservation, state_field="reservation_state")
+            if rsm.current_state != ReservationStateMachine.ReserveAborting:
+                nsi_exc = NsiException(
+                    InvalidTransition,
+                    rsm.current_state.name,
+                    {Variable.RESERVATION_STATE: rsm.current_state.value},
+                )
+                self.log.error("Cannot abort reservation", reason=nsi_exc.text)
+                # TODO: send a Service Exception to notify requester of nsi_exc
+                #
+                # self._send_service_exception(nsi_exc)
+            else:
+                try:
+                    #
+                    # FixME: If there is a Network Resource Manager that needs to be contacted
+                    #        to abort the reservation request then this is the place.
+                    #        If this is a recovered job then try to recover the reservation state
+                    #        from the NRM.
+                    #
+                    pass
+                except NsiException as nsi_exc:
+                    self.log.info("Reserve abort failed.", reason=nsi_exc.text)
+                    #
+                    # FIXME: Send NSI serviceException
+                    #
+                    # self._send_service_exception(session, nsi_exc)
+                except Exception as exc:
+                    self.log.exception("Unexpected error occurred.", reason=str(exc))
+                    #
+                    # FIXME: Send NSI serviceException
+                    #
+                    # nsi_exc = NsiException(GenericInternalError, str(exc))  # type: ignore[misc]
+                    # self._send_service_exception(session, nsi_exc)
+                else:
+                    rsm.reserve_abort_confirmed()
+                    self._send_reserve_abort_confirmed(session)
 
     @classmethod
-    def recover(cls) -> List[Job]:
-        """Recover ReserveAbortJobs."""
-        # Nothing to recover -> no-op
-        return super().recover()
+    def recover(cls: Type[ReserveAbortJob]) -> List[Job]:
+        """Recover ReserveAbortJob's that did not get to run before SuPA was terminated.
+
+        Returns:
+            List of ReserveAbortJob's that still need to be run.
+        """
+        from supa.db.session import db_session
+
+        with db_session() as session:
+            connection_ids: List[UUID] = list(
+                flatten(
```

Comment:
Not sure why you need the flatten? Perhaps change the query to reflect what you need.

Reply:
To turn a list of rows with connection_id's into a list of connection_id's.